### PR TITLE
change ai analytics

### PR DESCRIPTION
**Pull request by AntoineJac**
### Summary

This will allow to add a provide layer in the analytics in case several plugins are used 

Current model is:
```json
  "ai": {
    "meta": {
      "response_model": "mistral-tiny",
      "provider_name": "mistral",
      "request_model": "mistral-tiny"
    },
    "usage": {
      "prompt_tokens": 259,
      "completion_tokens": 274,
      "total_tokens": 533
    }
  }
```

New model if using several plugins:

```json
  "ai": {
    "mistral": {
      "request_total_tokens": 648,
      "number_of_instances": 1,
      "instances": [
        {
          "meta": {
            "request_model": "mistral-tiny",
            "plugin_id": "1c783122-0223-4a03-9f1d-c71d2baab01e",
            "response_model": "mistral-tiny",
            "provider_name": "mistral"
          },
          "usage": {
            "prompt_token": 392,
            "total_tokens": 648,
            "completion_token": 256
          }
        }
      ],
      "request_prompt_tokens": 392,
      "request_completion_tokens": 256
    },
    "azure": {
      "request_total_tokens": 145,
      "number_of_instances": 1,
      "instances": [
        {
          "meta": {
            "request_model": "gpt-35-turbo",
            "plugin_id": "5df193be-47a3-4f1b-8c37-37e31af0568b",
            "response_model": "gpt-35-turbo",
            "provider_name": "azure"
          },
          "usage": {
            "prompt_token": 89,
            "total_tokens": 145,
            "completion_token": 56
          }
        }
      ],
      "request_prompt_tokens": 89,
      "request_completion_tokens": 56
    },
    "cohere": {
      "request_total_tokens": 149,
      "number_of_instances": 1,
      "instances": [
        {
          "meta": {
            "request_model": "command",
            "plugin_id": "30a63ab9-03c6-447a-9b25-195eb0acaeb2",
            "response_model": "command",
            "provider_name": "cohere"
          },
          "usage": {
            "prompt_token": 78,
            "total_tokens": 149,
            "completion_token": 71
          }
        }
      ],
      "request_prompt_tokens": 78,
      "request_completion_tokens": 71
    }
  }
```

### Checklist

- [ ] The Pull Request has tests
- [ ] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

Fix #_[issue number]_

KAG-3759
KAG-4124

EDIT: format JSON snippets

### CODE REVIEW

**Comment by AntoineJac** [comment]
not new commit present

**Comment by AntoineJac** [comment]
By using the AI request and response transformers you will get following logs:

**When using same AI provider:**
`
"ai": {
    "cohere": {
      "instances": [
        {
          "meta": {
            "response_model": "command",
            "provider_name": "cohere",
            "request_model": "command"
          },
          "usage": {
            "completion_tokens": 33,
            "prompt_tokens": 129,
            "total_tokens": 162
          }
        },
        {
          "meta": {
            "response_model": "command",
            "provider_name": "cohere",
            "request_model": "command"
          },
          "usage": {
            "completion_tokens": 25,
            "prompt_tokens": 412,
            "total_tokens": 437
          }
        }
      ],
      "request_prompt_tokens": 541,
      "request_completion_tokens": 58,
      "request_total_tokens": 599,
      "number_of_instances": 2
    }
  }
 `

**When using different AI providers:**
`"ai": {
    "cohere": {
      "instances": [
        {
          "meta": {
            "response_model": "command",
            "provider_name": "cohere",
            "request_model": "command"
          },
          "usage": {
            "completion_tokens": 130,
            "prompt_tokens": 129,
            "total_tokens": 259
          }
        }
      ],
      "request_prompt_tokens": 129,
      "request_completion_tokens": 130,
      "request_total_tokens": 259,
      "number_of_instances": 1
    },
    "mistral": {
      "instances": [
        {
          "meta": {
            "response_model": "mistral-tiny",
            "provider_name": "mistral",
            "request_model": "mistral-tiny"
          },
          "usage": {
            "completion_tokens": 171,
            "prompt_tokens": 562,
            "total_tokens": 733
          }
        }
      ],
      "request_prompt_tokens": 562,
      "request_completion_tokens": 171,
      "request_total_tokens": 733,
      "number_of_instances": 1
    }
  }`

**Comment by subnetmarco** [comment]
Would it be possible to also store the `plugin_id` for each instance so that the user can point back to where it's being used?

File: kong/plugins/ai-proxy/handler.lua [hunk 1]
```diff
@@ -82,33 +82,72 @@ end
...
-  -- (kong.response.get_status() == 200) or (kong.ctx.plugin.ai_parser_error)
-
-  -- all errors MUST be checked and returned in header_filter
-  -- we should receive a replacement response body from the same thread
+  if kong.ctx.shared.skip_response_transformer then
+    local response_body, status
+    if kong.ctx.shared.parsed_response then
+      response_body = kong.ctx.shared.parsed_response
+    elseif kong.response.get_status() == 200 then
+      status, response_body = pcall( function()
```
Comment (jschmid1) [inline kong/plugins/ai-proxy/handler.lua:95]: Any specific reason you wrap `kong.service.get_raw_body` in a pcall?

File: spec/03-plugins/38-ai-proxy/02-openai_integration_spec.lua [hunk 2]
```diff
@@ -1,10 +1,63 @@
 local helpers = require "spec.helpers"
 local cjson = require "cjson"
 local pl_file = require "pl.file"
+local pl_stringx    = require "pl.stringx"
```
Comment (jschmid1) [inline spec/03-plugins/38-ai-proxy/02-openai_integration_spec.lua:4]:
```suggestion
local pl_stringx = require "pl.stringx"
```

File: spec/03-plugins/38-ai-proxy/02-openai_integration_spec.lua [hunk 3]
```diff
@@ -1,10 +1,63 @@
...
+
+      json = cjson.decode(data)
+    end)
+    .has_no_error("log file contains a valid JSON entry")
+
+  return json
+end
+
+local _EXPECTED_CHAT_STATS = {
+  openai = {
```
Comment (jschmid1) [inline spec/03-plugins/38-ai-proxy/02-openai_integration_spec.lua:38]: I think we want to cover all supported providers and not only openai, right?

File: spec/03-plugins/38-ai-proxy/02-openai_integration_spec.lua [hunk 4]
```diff
@@ -1,10 +1,63 @@
...
+      {
+        meta = {
+          plugin_id = '6e7c40f6-ce96-48e4-a366-d109c169e444',
+          provider_name = 'openai',
+          request_model = 'gpt-3.5-turbo',
+          response_model = 'gpt-3.5-turbo-0613',
+        },
+        usage = {
+          completion_token = 12,
+          prompt_token = 25,
```
Comment (jschmid1) [inline spec/03-plugins/38-ai-proxy/02-openai_integration_spec.lua:49]: In my tests I've seen this number to be inconsistent (with same input), which may result in flaky tests.

**Review by jschmid1 — CHANGES_REQUESTED** [review]

File: kong/llm/drivers/shared.lua [hunk 5]
```diff
@@ -247,61 +277,94 @@ function _M.pre_request(conf, request_table)
...
+    -- Get the current try count
+    local try_count = request_analytics_provider.number_of_instances
+
+    -- Set the model, response, and provider names in the current try context
+    current_try[log_entry_keys.META_CONTAINER][log_entry_keys.REQUEST_MODEL] = conf.model.name
+    current_try[log_entry_keys.META_CONTAINER][log_entry_keys.RESPONSE_MODEL] = response_object.model or conf.model.name
+    current_try[log_entry_keys.META_CONTAINER][log_entry_keys.PROVIDER_NAME] = conf.model.provider
+    current_try[log_entry_keys.META_CONTAINER][log_entry_keys.PLUGIN_ID] = conf.__plugin_id
+
+    -- Capture openai-format usage stats from the transformed response body
```
[the remark this answers is not in the document]
Comment (tysoekong) [reply kong/llm/drivers/shared.lua:338]: They all output "prompt tokens" and "completion tokens", and our code normalizes the response data to make sure they're captured for all providers.

File: kong/plugins/ai-proxy/handler.lua (hunk 1 above)
Comment (tysoekong) [reply kong/plugins/ai-proxy/handler.lua:95]: @AntoineJac  ?

File: kong/llm/drivers/shared.lua [hunk 6]
```diff
@@ -247,61 +277,94 @@ function _M.pre_request(conf, request_table)
...
+        instances = {},
       }
     end
 
-    -- this captures the openai-format usage stats from the transformed response body
+    -- Increment the number of instances
+    request_analytics_provider.number_of_instances = request_analytics_provider.number_of_instances + 1
+    
+    -- Get the current try count
+    local try_count = request_analytics_provider.number_of_instances
```
[the remark this answers is not in the document]
Comment (tysoekong) [reply kong/llm/drivers/shared.lua:330]: @AntoineJac

File: kong/llm/drivers/shared.lua [hunk 7]
```diff
@@ -247,61 +277,94 @@ function _M.pre_request(conf, request_table)
...
     if not request_analytics then
-      request_analytics = {
-        prompt_tokens = 0,
-        completion_tokens = 0,
-        total_tokens = 0,
+      request_analytics = {}
+    end
+
+    -- check if we already have analytics for this provider
+    local request_analytics_provider = request_analytics[conf.model.provider]
```
[the remark this answers is not in the document]
Comment (AntoineJac) [reply kong/llm/drivers/shared.lua]: Fix in last commit

File: kong/plugins/ai-proxy/handler.lua (hunk 1 above)
Comment (AntoineJac) [reply kong/plugins/ai-proxy/handler.lua]: Fix in last commit

File: kong/llm/drivers/shared.lua (hunk 6 above)
[the remark this answers is not in the document]
Comment (AntoineJac) [reply kong/llm/drivers/shared.lua]: Fixed in last commit

File: kong/llm/drivers/shared.lua [hunk 8]
```diff
@@ -26,6 +34,28 @@ _M.streaming_has_token_counts = {
   ["cohere"] = true,
   ["llama2"] = true,
 }
+local function split_table_key(obj)
```
[the remark this answers is not in the document]
Comment (AntoineJac) [reply kong/llm/drivers/shared.lua:42]: @jschmid1  where should I add this function?
Should I create a `31-llm_shared_spec.lua ` under https://github.com/Kong/kong-ee/tree/master/spec/01-unit?

File: spec/03-plugins/38-ai-proxy/02-openai_integration_spec.lua (hunk 3 above)
Comment (AntoineJac) [reply spec/03-plugins/38-ai-proxy/02-openai_integration_spec.lua:38]: Same that below  "The response is static" and transformed by the ai plugins for all providers.

**Comment by AntoineJac** [comment]
It seems we have resolved all the points. Are we ok to push this PR? @jschmid1 
Thanks

**Review by jschmid1 — APPROVED** [review]

**Comment by AntoineJac** [comment]
@jschmid1  thanks could you please cherry pick and merge to EE? Thanks

**Comment by dndx** [comment]
@jschmid1 This squashed commit is missing the component in the scope, please watch out next time :)

**Pull request merged**